lib/services/access_request.go

```diff
@@ -245,53 +344,246 @@ func NewAccessRequest(user string, roles ...string) (AccessRequest, error) {
 	return &req, nil
 }
 
+func (c AccessRequestConditions) GetTraitMappings() TraitMappingSet {
+	tm := make([]TraitMapping, 0, len(c.ClaimsToRoles))
+	for _, mapping := range c.ClaimsToRoles {
+		tm = append(tm, TraitMapping{
+			Trait: mapping.Claim,
+			Value: mapping.Value,
+			Roles: mapping.Roles,
+		})
+	}
+	return TraitMappingSet(tm)
+}
+
 type UserAndRoleGetter interface {
 	UserGetter
 	RoleGetter
+	GetRoles() ([]Role, error)
 }
 
-func ValidateAccessRequest(getter UserAndRoleGetter, req AccessRequest) error {
-	user, err := getter.GetUser(req.GetUser(), false)
+// appendRoleMatchers constructs all role matchers for a given
+// AccessRequestConditions instance and appends them to the
+// supplied matcher slice.
+func appendRoleMatchers(matchers []parse.Matcher, conditions AccessRequestConditions, traits map[string][]string) ([]parse.Matcher, error) {
```

Comment:
style nit (optional), not sure how much `append` simplifies the logic, if you have `collectMatchers` that has no side effects you can always do:

```go
matchers = append(matchers, collectMatchers(allow, traits)...)
```